server/api_archive.go

```diff
@@ -0,0 +1,126 @@
+package server
+
+// This is a low-level archiver which zips and sends GCA server persistent file
+// data using a POST command. This is intended to be used by an external service
+// to provide archival snapshots of raw data in the system. Only public data is
+// provided, with private keys zeroed before sending.
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+
+	"github.com/glowlabs-org/gca-backend/glow"
+)
+
+var (
+	ApiArchiveRateLimiter *glow.RateLimiter // Provide a rate limiter for this API.
+)
+
+func init() {
+	ApiArchiveRateLimiter = glow.NewRateLimiter(apiArchiveLimit, apiArchiveRate)
+}
+
+// ArchiveHandler provides the POST /archive api endpoint. It returns
+// uninterpreted file data intended to be used by an external service
+// to produce an archive.
+func (gcas *GCAServer) ArchiveHandler(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodPost {
+		http.Error(w, "Only POST method is allowed", http.StatusMethodNotAllowed)
+		return
+	}
+
+	if r.ContentLength != 0 {
+		http.Error(w, "Request should not have a body", http.StatusBadRequest)
+		return
+	}
+
+	if !ApiArchiveRateLimiter.Allow() {
+		http.Error(w, "Too many requests", http.StatusTooManyRequests)
+		return
+	}
+
+	buffer := new(bytes.Buffer)
+	zipw := zip.NewWriter(buffer)
+
+	files := []string{"server.keys", "gcaTempPubKey.dat", "gcaPubKey.dat", "equipment-authorizations.dat", "allDeviceStats.dat", "equipment-reports.dat"}
+
+	for _, name := range files {
+		err := gcas.addFile(name, zipw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("Error zipping file %v: %v", name, err), http.StatusInternalServerError)
+			return
+		}
+	}
+
+	if err := zipw.Close(); err != nil {
+		if _, err := w.Write(buffer.Bytes()); err != nil {
+			http.Error(w, "Failed to close zip writer", http.StatusInternalServerError)
+			return
+		}
+	}
+
+	w.Header().Set("Content-Type", "application/zip")
+
+	if _, err := w.Write(buffer.Bytes()); err != nil {
+		http.Error(w, "Failed to write archive response", http.StatusInternalServerError)
+	}
+}
+
+func (gcas *GCAServer) addFile(name string, zipw *zip.Writer) error {
+	// To avoid races with any components writing to the file, lock up here.
+	gcas.mu.Lock()
+	defer gcas.mu.Unlock()
+
+	path := filepath.Join(gcas.baseDir, name)
+
+	file, err := os.Open(path)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	fileInfo, err := file.Stat()
+	if err != nil {
+		return err
+	}
+
+	header, err := zip.FileInfoHeader(fileInfo)
+	if err != nil {
+		return err
+	}
+
+	header.Method = zip.Deflate
+
+	writer, err := zipw.CreateHeader(header)
+	if err != nil {
+		return err
+	}
+
+	if name == "server.keys" {
+		// Special case, this file contains private data, so we need to sanitize it.
```

Comment:
This is probably going to cause confusion. The unzipped archive is going to look like a server that is totally ready to go, but then when someone tries to run the archive they are going to get errors because the server is not going to be able to sign anything. Instead you should probably save an entirely separate filename `server.pubkey` and then also add a README to the archive that explains the server private keys are not put into the archive, and therefore the archive can't be naively used to stand up a new server but instead additional work is needed to get functional keys in place.

Reply:
Will rename this file the server_public_key.dat and include only the public key to make clearer